vm/ubpf_vm.c

```diff
@@ -1259,6 +1259,85 @@ ubpf_exec_ex(
             // Because we have already validated, we can assume that the type code is
             // valid.
             break;
+        case EBPF_OP_ATOMIC_STORE: {
+            BOUNDS_CHECK_STORE(8);
+            bool fetch = inst.imm & EBPF_ATOMIC_OP_FETCH;
+            volatile uint64_t* destination = (volatile uint64_t*)(reg[inst.dst] + inst.offset);
+            uint64_t value = reg[inst.src];
+            uint64_t result;
+            switch (inst.imm & EBPF_ALU_OP_MASK) {
+            case EBPF_ALU_OP_ADD:
+                result = UBPF_ATOMIC_ADD_FETCH(destination, value);
+                break;
+            case EBPF_ALU_OP_OR:
+                result = UBPF_ATOMIC_OR_FETCH(destination, value);
+                break;
+            case EBPF_ALU_OP_AND:
+                result = UBPF_ATOMIC_AND_FETCH(destination, value);
+                break;
+            case EBPF_ALU_OP_XOR:
+                result = UBPF_ATOMIC_XOR_FETCH(destination, value);
+                break;
+            case (EBPF_ATOMIC_OP_XCHG & ~EBPF_ATOMIC_OP_FETCH):
+                result = UBPF_ATOMIC_EXCHANGE(destination, value);
+                break;
+            case (EBPF_ATOMIC_OP_CMPXCHG & ~EBPF_ATOMIC_OP_FETCH):
+                result = UBPF_ATOMIC_COMPARE_EXCHANGE(destination, value, reg[0]);
+                break;
+            default:
+                vm->error_printf(stderr, "Error: unknown atomic opcode %d at PC %d\n", inst.imm, cur_pc);
+                return_value = -1;
+                goto cleanup;
+            }
+            if (fetch) {
+                if (inst.imm != EBPF_ATOMIC_OP_CMPXCHG) {
+                    reg[inst.src] = result;
+                }
+                else {
+                    reg[0] = result;
+                }
+            }
+        } break;
+
+        case EBPF_OP_ATOMIC32_STORE: {
+            BOUNDS_CHECK_STORE(4);
+            bool fetch = inst.imm & EBPF_ATOMIC_OP_FETCH;
+            volatile uint32_t* destination = (volatile uint32_t*)(reg[inst.dst] + inst.offset);
```

Comment:
I am not 100% sure that this matters or whether it violates the semantics of the operation, but ...

Here were are using the value of a register (with the identifier `inst.dst`) to calculate the offset of the memory with which we are going to work atomically. However, there is a time gap between when we calculate `destination` using that value and the time that we actually read from it. What if the value of the register with id `inst.dst` changes in that intermediate time?

Again, I am not 100% sure whether it is a problem and/or the semantics that we should guarantee?

Reply:
Register value is stored on the stack and is single threaded, hence no race condition.